cli/lsp/documents.rs

```diff
@@ -261,49 +361,113 @@ impl Document {
       }
     }
     let content = Arc::new(content);
-    if self
+    let maybe_module = if self
+      .inner
       .maybe_language_id
       .as_ref()
       .map(|li| li.is_diagnosable())
       .unwrap_or(false)
     {
       let maybe_headers = self
+        .inner
         .maybe_language_id
         .as_ref()
         .map(|li| li.as_headers())
         .flatten();
       let parser = SourceParser::default();
-      self.maybe_module = Some(deno_graph::parse_module(
-        &self.specifier,
+      Some(deno_graph::parse_module(
+        &self.inner.specifier,
         maybe_headers,
         content.clone(),
         maybe_resolver,
         Some(&parser),
-      ));
+      ))
     } else {
-      self.maybe_module = None;
-    }
-    self.source = SourceTextInfo::new(content);
-    self.line_index = if index_valid == IndexValid::All {
+      None
+    };
+    let source = SourceTextInfo::new(content);
+    let line_index = if index_valid == IndexValid::All {
       line_index
     } else {
-      Arc::new(LineIndex::new(self.source.text_str()))
+      Arc::new(LineIndex::new(source.text_str()))
     };
-    self.maybe_lsp_version = Some(version);
-    self.maybe_navigation_tree = None;
-    Ok(())
+    Ok(Document {
+      inner: Arc::new(DocumentInner {
+        specifier: self.inner.specifier.clone(),
+        version: self.inner.version.clone(),
+        text_info: source,
+        line_index,
+        maybe_module,
+        maybe_language_id: self.inner.maybe_language_id.clone(),
+        maybe_warning: self.inner.maybe_warning.clone(),
+        maybe_lsp_version: Some(version),
+        maybe_navigation_tree: None,
+      }),
+    })
   }
 
-  fn close(&mut self) {
-    self.maybe_lsp_version = None;
-    self.maybe_language_id = None;
+  fn with_closed(&self) -> Document {
```

Comment:
I wasn't so concerned about the `with_change`, but this seems totally arbitrary and "wasteful".  We care cloning everything to set two properties to `None`.

Other than a bit of readability, I am not sure there really is a compelling case here.  I think we need to get some wider opinions about this, like from @lucacasonato.

Reply:
It's because the document structs are now immutable and cheaply cloned. This will allow us to get "snapshots" of the LSP state for actual multiple messages in the future. Right now the "`StateSnapshot`" is a lie (see how it holds documents, which is an `Arc<Mutex<_>>`). It will also allow us to use a document in places where previously we would have to pass the entire documents collection and we can move these documents into other kinds of collections and still have this functionality rather than it being tied to this specific documents collection API.

I've made the statement look a little less verbose by using a spread operator and clone the inner.

I don't believe this is wasteful. There isn't anything in there that seems heavy to clone. Could you elaborate?

Reply:
The "heaviest" clone I could find in here is the `deno_graph::Module` in the `maybe_module` slot. Cloning that will clone the entire source text of the module. In the grand scope of things, I think the memory pressure vs. immutability tradeoff is worth it here, because the immutable documents will make it easier to implement concurrent LSP calls (and thus cancellations!)

Reply:
@lucacasonato this is not a big deal, because source code of the module is already stored in an `Arc` https://docs.rs/deno_graph/0.11.1/deno_graph/struct.Module.html

Reply:
I mentioned below, I think if we move NavTree out of document and the lsp version, the document truly becomes immutable, and we wouldn't clone on any other modification.

Reply:
@kitsonk I think we should still store that in the `Document` because moving it out would mean consulting some other source of truth which might not align with `Document`. Maybe we should move all the properties that won't change into another sub struct and throw that in an `Arc`? I'm not sure it's worth it. Eventually we'll be able to remove the navigation tree.

I think with this approach we'll eventually be able to create real snapshots of the documents at the start of an LSP request and not need to check any data behind a lock that could be mutated by another thread (side note: I think I have an idea for how to handle closed files without running into race conditions... need to explore it a bit).